src/analysis/processing/qgsalgorithmkmeansclustering.cpp

```diff
@@ -0,0 +1,371 @@
+  int i = 0;
+  int n = 0;
+  QgsFeature feat;
+
+  std::vector< Feature > clusterFeatures;
+  QgsFeatureIterator features = source->getFeatures( QgsFeatureRequest().setSubsetOfAttributes( QgsAttributeList() ) );
+  QHash< QgsFeatureId, int > idToObj;
+  while ( features.nextFeature( feat ) )
+  {
+    i++;
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+
+    feedback->setProgress( i * step );
+    if ( !feat.hasGeometry() )
+      continue;
+
+    n++;
+
+    QgsPointXY point;
+    if ( QgsWkbTypes::flatType( feat.geometry().wkbType() ) == QgsWkbTypes::Point )
+      point = QgsPointXY( *qgsgeometry_cast< const QgsPoint * >( feat.geometry().constGet() ) );
+    else
+    {
+      QgsGeometry centroid = feat.geometry().centroid();
+      point = QgsPointXY( *qgsgeometry_cast< const QgsPoint * >( centroid.constGet() ) );
+    }
+
+    idToObj.insert( feat.id(), clusterFeatures.size() );
+    clusterFeatures.emplace_back( Feature( point ) );
+  }
+
+  if ( n < k )
+  {
+    feedback->reportError( QObject::tr( "Number of geometries is less than the number of clusters requested, not all clusters will get data" ) );
+    k = n;
+  }
+
+  if ( k > 1 )
+  {
+    feedback->pushInfo( QObject::tr( "Calculating clusters" ) );
+
+    // cluster centers
+    std::vector< QgsPointXY > centers( k );
+
+    initClusters( clusterFeatures, centers, k, feedback );
+    calculateKMeans( clusterFeatures, centers, k, feedback );
+  }
+
+  features = source->getFeatures();
+  i = 0;
+  while ( features.nextFeature( feat ) )
+  {
+    i++;
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+
+    feedback->setProgress( 50 + i * step );
+    QgsAttributes attr = feat.attributes();
+    if ( !feat.hasGeometry() )
+    {
+      attr << QVariant();
+    }
+    else if ( k <= 1 )
+    {
+      attr << 0;
+    }
+    else if ( !idToObj.contains( feat.id() ) )
+    {
+      attr << QVariant();
+    }
+    else
+    {
+      attr << clusterFeatures[ idToObj.value( feat.id() ) ].cluster;
+    }
+    feat.setAttributes( attr );
+    sink->addFeature( feat, QgsFeatureSink::FastInsert );
+  }
+
+  QVariantMap outputs;
+  outputs.insert( QStringLiteral( "OUTPUT" ), dest );
+  return outputs;
+}
+
+// ported from https://github.com/postgis/postgis/blob/svn-trunk/liblwgeom/lwkmeans.c
+
+void QgsKMeansClusteringAlgorithm::initClusters( std::vector<Feature> &points, std::vector<QgsPointXY> &centers, const int k, QgsProcessingFeedback *feedback )
+{
+  ulong n = points.size();
+  if ( n == 0 )
+    return;
+
+  if ( n == 1 )
+  {
+    for ( int i = 0; i < k; i++ )
+      centers[ i ] = points[ 0 ].point;
+    return;
+  }
+
+  long duplicateCount = 1;
+  // initially scan for two most distance points from each other, p1 and p2
+  ulong p1 = 0;
+  ulong p2 = 0;
+  double distanceP1 = 0;
+  double distanceP2 = 0;
+  double maxDistance = -1;
+  for ( ulong i = 1; i < n; i++ )
+  {
+    distanceP1 = points[i].point.sqrDist( points[p1].point );
```

Comment:
I'm facing the "problem" to cluster with not-euclidean or not geodetic disntances, eg. cluster on clusterised data to narrow or groups. The main problems of the available clustering methods is that they do not accept a method to measure the not-euclidean distances.
would be fine to add a parameter to accept a function or a callable that calc the dinstance.
BTW nice to have clustering alg in QGIS without need to use a backend. Tnx

Reply:
@luipir the problem is that this algorithm will not work correctly with arbitrary distance measures. (And even with Euclidean distances it is not guaranteed to find optimal solution because it is a heuristics - but usually it works fine)

Reply:
@luipir

As @wonder-sk noted there's lots of potential issues here. I think for now what I'll do is rename this alg to "K means clustering (Cartesian distance)", so that in future we can add a ellipsoidal distance one and an attribute/expression based one.

I'll also refactor this method so that we can avoid duplicate code if/when these are created

Reply:
great, tnx